benchmarks/prescribe_dike_injection/prescribed_dike_injection.cc

```diff
@@ -0,0 +1,571 @@
+      *prescribed_dilation = (this->get_parameters().enable_prescribed_dilation)
+                              ? out.template get_additional_output<MaterialModel::PrescribedPlasticDilation<dim> >()
+                              : nullptr;
+      ReactionRateOutputs<dim>
+      *reaction_rate_out = (this->get_parameters().use_operator_splitting)
+                            ? out.template get_additional_output<MaterialModel::ReactionRateOutputs<dim>>()
+                            : nullptr;
+      // Initiallize reaction_rates to 0.0.
+      if (reaction_rate_out != nullptr)
+        for (auto &row : reaction_rate_out->reaction_rates)
+          std::fill(row.begin(), row.end(), 0.0);
+
+      // When calculating other properties such as viscosity, we need to 
+      // correct for the effect of injection on the strain rate and thus
+      // the deviatoric strain rate.
+      MaterialModel::MaterialModelInputs<dim> in_corrected_strainrate (in);
+
+      // Strore dike injection rate for each evaluation point
+      std::vector<double> dike_injection_rate(in.n_evaluation_points());
+
+      for (unsigned int i=0; i < in.n_evaluation_points(); ++i)
+        {
+          // First find the location of the dike which is either randomly
+          // generated or prescribed by the user.
+          // Then give the dike_injection_rate to the dike points.
+          if (enable_random_dike_generation)
+            {
+              // Note: when the dike is generated randomly, the prescribed
+              // injection rate in the 'Dike injection function' should be
+              // only time dependent and independent of the xyz-coordinate.
+              const double point_depth = this->get_geometry_model().depth(in.position[i]);
+
+              // Find the randomly generated dike location
+              if (in.position[i][0] >= x_dike_left_boundary
+                  && in.position[i][0] <= x_dike_right_boundary
+                  && in.temperature[i] <= T_bottom_dike 
+                  && point_depth >= std::max(top_depth_random_dike, 0.0)
+                  && this->get_timestep_number() > 0)
+                dike_injection_rate[i] = this->convert_output_to_years()
+                                         ? injection_function.value(in.position[i]) / year_in_seconds
+                                         : injection_function.value(in.position[i]);
+              else
+                dike_injection_rate[i] = 0.0;
+
+              // Dike injection effect removal
+              // Note that the correction starts from Timestep 1.
+              // Ensure the current cell is located within the dike area.
+              if (dike_injection_rate[i] > 0.0
+                  && this->get_timestep_number() > 0
+                  && in.current_cell.state() == IteratorState::valid
+                  && in.current_cell->center()[0] >= x_dike_left_boundary
+                  && in.current_cell->center()[0] <= x_dike_right_boundary)
```

Comment:
I'm also wondering if they are correct. The dike width can be set by the user to anything, which means the center of the dike doesn't necessarily need to lie within the left and right boundary. It will always be larger than the left boundary, because this is aligned with the left boundary of the highest resolution cell, but the right boundary is not necessarily aligned with any cell boundary. See line 112.

Reply:
> Are these conditions necessary? Will the dike_injection_rate not be zero if the cell center doesn't lie within the boundaries?

Yes, these conditions are set to solve visualization problems: In Paraview, the right vertices of the dike cell overlap with the left vertices of its left neighbor, causing the latter to show the high strain rate caused by the dike.

Reply:
> I'm also wondering if they are correct. The dike width can be set by the user to anything, which means the center of the dike doesn't necessarily need to lie within the left and right boundary. It will always be larger than the left boundary, because this is aligned with the left boundary of the highest resolution cell, but the right boundary is not necessarily aligned with any cell boundary. See line 112.

Yes. So it is better to set the width as the same as the maximum x-resolution or an integer multiple thereof. I added these explanations to the plugin.